asmcomp/amd64/emit.mlp

```diff
@@ -734,6 +734,23 @@ let emit_instr fallthrough i =
   | Lop(Iintop(Ilsl | Ilsr | Iasr as op)) ->
       (* We have i.arg.(0) = i.res.(0) and i.arg.(1) = %rcx *)
       instr_for_intop op cl (res i 0)
+  | Lop (Iintop(Iclz non_zero_input)) ->
+      if non_zero_input then
+        (* No need to handle that bsr is undefind on 0 input. *)
+        (I.bsr (arg i 0) (res i 0);
+         I.xor (int 63) (res i 0))
```

Comment:
It wasn't obvious at all to me! That's the code generated by gcc for _builtin_clz. The main reason I think is that there seem to be no "reverse subtract" instruction. The first argument to subtract has to be a register. To implement "64 -bsr" using subtract, we would first need to negate bsr result (hence the leftover neg instruction you noticed above) and then add 64.

Reply:
Right, but `64 - x` is not the same as `63 xor x`, and one needs to understand why the latter is correct, which is not in line with was I understand from the spec. (I'm not saying this is wrong, esp. since you have tests that would have caught it.)

Reply:
BSR returns a result between 0 (when there are 63 leading zeros) and 63 (when there is no leading zero) so it's not `64 - x` that you need, but `63 - x`, which is of course equal to `63 xor x`.